### ISSUE TITLE
Performance benchmarks/metrics for iot-edge?

### ISSUE THREAD

**Issue by malichishti** [Azure/iot-edge#381]
Hi,

I would like to know if there exists any performance testing results and/or benchmarks around the iot edge message bus? 
How many events it can handle/ingest per second?
What if one module is running slow in the pipeline? how long will the events be buffered if there is any buffering mechanism?
What happens if the CPU memory if over flown due to large number of messages in buffer/queue?

Regards,
Ali

**Comment by darobs** [comment]
We do have Performance/stress tests built into the E2E tests. They're simple tests which one module produces messages as fast as possible, and a second module measures how well we receive them.  There are options on the producer module to tune the output.

If you build the project with the "--run-e2e-tests" these tests will be built.  The code can be found in `core/tests/performance_e2e`

This is an example output, from the last unix build (an x64 VM), 5 second run:

> 16: Duration (ms): 4859
> 16: Messages received: 356609
> 16: Non-Conforming Messages: 0
> 16: Message Latency (average microseconds): 20
> 16: Message Latency (max microseconds): 8561
> 16: Devices Discovered: 1
> 16: Device: device1
> 16: Message count: 356609
> 16: Out of Sequence Count: 0
> 16: Messages Lost: 0
> 16: Suceeded.

We've discovered that nanomsg will drop messages if its queues fill up. Running these performance tests over longer periods of time generally result in dropped messages, as the producer module does not have any flow control built into it.

**Comment by malichishti** [comment]
So its correct to say that there is no congestion control built into the sdk?

Regards,
Ali

**Comment by darobs** [comment]
We are relying on nanomsg for message delivery, and while it does have some flow control, only one topology (req resp) has guaranteed delivery, and that is not the topology we're using (we're using pub sub).  So, once queues start to fill, messages may drop. 

I was able to reduce or eliminate (depending on platform) dropped messages by setting socket options to expand the queue sizes, but it was not a delivery guarantee, so we opted not to change the default settings.
